overlord/devicestate/users.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, err
+	}
+	return u, nil
+}
+
+func getUserDetailsFromStore(st *state.State, theStore snapstate.StoreService, email string) (string, *osutil.AddUserOptions, error) {
+	st.Unlock()
+	defer st.Lock()
+
+	v, err := theStore.UserInfo(email)
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot create user %q: %s", email, err)
+	}
+	if len(v.SSHKeys) == 0 {
+		return "", nil, fmt.Errorf("cannot create user for %q: no ssh keys found", email)
+	}
+
+	// Amend information where the key came from to ensure it can
+	// be update/replaced later
+	for i, k := range v.SSHKeys {
+		v.SSHKeys[i] = fmt.Sprintf(`%s # snapd {"origin":"store","email":%q}`, k, email)
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, v.OpenIDIdentifier)
+	opts := &osutil.AddUserOptions{
+		SSHKeys: v.SSHKeys,
+		Gecos:   gecos,
+	}
+	return v.Username, opts, nil
+}
+
+func createKnownSystemUser(state *state.State, userAssertion *asserts.SystemUser, assertDb asserts.RODatabase, model *asserts.Model, serial *asserts.Serial, sudoer bool) (*CreatedUser, error) {
+	email := userAssertion.Email()
+	// we need to use getUserDetailsFromAssertion as this verifies
+	// the assertion against the current brand/model/time
+	username, addUserOpts, err := getUserDetailsFromAssertion(assertDb, model, serial, email)
+	if err != nil {
+		logger.Noticef("ignoring system-user assertion for %q: %s", email, err)
+		return nil, nil
+	}
+
+	// ignore already existing users
+	if _, err := userLookup(username); err == nil {
+		return nil, nil
+	}
+
+	addUserOpts.Sudoer = sudoer
+	return addUser(state, username, email, addUserOpts)
+}
+
+func createAllKnownSystemUsers(state *state.State, assertDb asserts.RODatabase, model *asserts.Model, serial *asserts.Serial, sudoer bool) ([]*CreatedUser, error) {
+	headers := map[string]string{
+		"brand-id": model.BrandID(),
+	}
+
+	assertions, err := assertDb.FindMany(asserts.SystemUserType, headers)
+	if err != nil && !asserts.IsNotFound(err) {
+		return nil, &UserError{Err: fmt.Errorf("cannot find system-user assertion: %s", err)}
+	}
+
+	var createdUsers []*CreatedUser
+	for _, as := range assertions {
+		createdUser, err := createKnownSystemUser(state, as.(*asserts.SystemUser), assertDb, model, serial, sudoer)
+		if err != nil {
+			return nil, err
+		}
+		if createdUser == nil {
+			continue
+		}
+		createdUsers = append(createdUsers, createdUser)
+	}
+
+	return createdUsers, nil
+}
+
+func getUserDetailsFromAssertion(assertDb asserts.RODatabase, modelAs *asserts.Model, serialAs *asserts.Serial, email string) (string, *osutil.AddUserOptions, error) {
+	errorPrefix := fmt.Sprintf("cannot add system-user %q: ", email)
+
+	brandID := modelAs.BrandID()
+	series := modelAs.Series()
+	model := modelAs.Model()
+
+	a, err := assertDb.Find(asserts.SystemUserType, map[string]string{
+		"brand-id": brandID,
+		"email":    email,
+	})
+	if err != nil {
+		return "", nil, fmt.Errorf(errorPrefix+"%v", err)
```

Comment:
let's try to never concatenate strings inside a printf/errorf function, especially if the concatenated string itself might contain user data. Maybe it's impossible, but if the `email` variable contains a `%s`, then we are running a security risk here, where our stack might get compromised.

Let's use
```suggestion
		return "", nil, fmt.Errorf("%s%v", errorPrefix, err)
```
instead.

Reply:
There are quite a few of these below, too.

Reply:
Really nice catch! I'll get these fixed